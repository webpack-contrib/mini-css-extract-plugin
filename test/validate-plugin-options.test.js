import MiniCssExtractPlugin from '../src';

describe('validate options', () => {
  const tests = {
    filename: {
      success: ['[name].css'],
      failure: [true],
    },
    chunkFilename: {
      success: ['[id].css'],
      failure: [true],
    },
    moduleFilename: {
      success: [({ name }) => `${name.replace('/js/', '/css/')}.css`],
      failure: [true],
    },
    ignoreOrder: {
      success: [true, false],
      failure: [1],
    },
    unknown: {
      success: [],
      // TODO failed in next release
      // failure: [1, true, false, 'test', /test/, [], {}, { foo: 'bar' }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      let error;

      try {
        // eslint-disable-next-line no-new
        new MiniCssExtractPlugin({ [key]: value });
      } catch (errorFromPlugin) {
        if (errorFromPlugin.name !== 'ValidationError') {
          throw errorFromPlugin;
        }

        error = errorFromPlugin;
      } finally {
        if (type === 'success') {
          expect(error).toBeUndefined();
        } else if (type === 'failure') {
          expect(() => {
            throw error;
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
