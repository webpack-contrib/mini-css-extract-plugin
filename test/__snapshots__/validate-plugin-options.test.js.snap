// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "chunkFilename" option with "true" value 1`] = `
"Mini CSS Extract Plugin Invalid Options

options.chunkFilename should be string
"
`;

exports[`validate options should throw an error on the "filename" option with "true" value 1`] = `
"Mini CSS Extract Plugin Invalid Options

options.filename should be string
"
`;

exports[`validate options should throw an error on the "ignoreOrder" option with "1" value 1`] = `
"Mini CSS Extract Plugin Invalid Options

options.ignoreOrder should be boolean
"
`;

exports[`validate options should throw an error on the "moduleFilename" option with "true" value 1`] = `
"Mini CSS Extract Plugin Invalid Options

options.moduleFilename should pass \\"instanceof\\" keyword validation
"
`;
